### ISSUE TITLE
Feature suggestion: Animated cursor

### ISSUE THREAD

**Issue by Asheq** [kovidgoyal/kitty#2460]
https://github.com/Kethku/neovide/blob/master/assets/AnimatedCursor.gif

The above gif is from a neovim front-end and demonstrates an animated cursor. Using it is actually quite a joy, and useful because it's easier for you and onlookers to track the cursor position.

@kovidgoyal Just wondering if you would be willing to consider something like this for kitty (as an option that can be disabled or enabled, of course). Thanks!

**Comment by ctrlcctrlv** [comment]
Wow, cool. I support this idea. I'm already thinking what kind of clever shader can do this.

**Comment by Luflosi** [comment]
Looks cool but I have no idea how this might be implemented, this is probably not that simple.

**Comment by kovidgoyal** [comment]
Seems somewhat headache inducing to me, but am fine with an option for it that defaults to off. To implement it you would basically need to use the timers facility to implement the animation, similar to how it is done for blinking cursors. 

If someone else wants to work on it feel free to post if you need help and I will provide assistance with getting started.

**Comment by egmontkob** [comment]
Let's note that in terminals, due to the very nature of the architecture, this can easily result in faulty jumps, i.e. jumps to intermediate positions where the contents are updated but the cursor is not meant to rest.

This can be mitigated for cooperating apps that either hide the cursor while repainting, or use the protocol extension for atomic updates (I can't recall if Kitty already supports it, or we just discussed it, sorry).

Still, the terminal needs to figure out what to do if a cursor animation is in progress and in the mean time the cursor moves to a new location, e.g. should it complete the pending animation first, or should it change direction immediately, or maybe should it consider the pending animation an unintended one, cancel it, and restart a new animation from the previous starting point (maybe if this new animation is not more than x% complete). It's heuristics, and no matter what do you, there will be cases when the user sees undesired glitches.

Also there are many more kinds of apps to consider, not just text editors; e.g. what would the cursor do while editing a multi-line commandline, cat'ing a large file, or using mc, less, top, alsamixer, tmux with split panes etc...

Would be interesting to see a prototype, see all these in live action ;) Without that, I can't really foresee if this feature would be more on the "super useful, with very few glitches" side or on the "quite useless because of way too many glitches" side. If the latter, you might consider introducing an escape sequence, so that it can be enabled for specific apps only.

**Comment by kovidgoyal** [comment]
On Mon, Mar 23, 2020 at 02:29:08AM -0700, egmontkob wrote:
> Let's note that in terminals, due to the very nature of the architecture, this can easily result in faulty jumps, i.e. jumps to intermediate positions where the contents are updated but the cursor is not meant to rest.

Why? I dont follow. The actual cursor position and movement semantics
remain the same, this is pure eye candy. So to a terminal program the
cursor will behave as it always has. To a user looking at the screen
cursor movements will be "animated". To illustrate:

cursor at (1, 1) cursor moves to (5, 7)

as soon as the escape sequence to move to (5, 7) is received, the actual
cursor moves to (5, 7), however, on screen, the user sees an animated
motion over a few frames. If another movement command is received while
the animation is in progress, the animation is aborted or re targeted to
the new destination.

> Would be interesting to see a prototype, see all these in live action ;) Without that, I can't really foresee if this feature would be more on the "super useful, with very few glitches" side or on the "quite useless because of way too many glitches" side. If the latter, you might consider introducing an escape sequence, so that it can be enabled for specific apps only.

TO me this is headache inducing, I really dislike stuff jumping around
on screen, so am not a fan of it. And I agree, one would need to see
how it works in practice to decide what level of granularity of control
is needed. Options are:

1) Turned onoff globally via a terminal config setting
2) Turned on/off globally via an escape code
3) Turned on only for individual cursor movements, so you would have an
escape code that specifies a cursor movement and optionally whether it
should be animated.

**Comment by egmontkob** [comment]
> this is pure eye candy

Pure eye candy, but not as intended.

A graphical application can do the animation as desired, because the concept of visual cursor and the concept of updating parts of the screen are independent. In terminals, updating the screen can pretty much only happen via the cursor, so the two concepts are heavily coupled.

Let's say, the user uses vim and presses a key that causes the cursor to jump from (1,1) to (5,7). What in fact happens behind the scenes is that the cursor moves to (60,24) (or wherever), the cursor position as it is printed in vim's status bar gets updated (moving the cursor forward through (61,24), (62,24) etc.), and finally it jumps to (5,7).

More complex changes might easily cause the cursor to paste through much larger portions of the screen, or even literally the entire screen.

If the cursor flashes up for a visual frame in one of these intermittent locations, it's probably not noticeable, or at least not distracting. If it's animated, even if simply from (1,1) to (60,24) and then from (60,24) to (5,7), it is a heavily noticeable undesired glitch – which is inevitable to occur every once in a while, due to the timing of events, network lag etc.

> TO me this is headache inducing, I really dislike stuff jumping around on screen

I fully agree with you; by looking at the animgif I'm sure I would want this feature to be disabled. However, this is not the aspect I meant to comment on, since it's a user preference. I meant to point out the technical problems, in case one wishes to have this feature.

**Comment by kovidgoyal** [comment]
On Mon, Mar 23, 2020 at 03:29:55AM -0700, egmontkob wrote:
> > this is pure eye candy
> 
> Pure eye candy, but not as intended.
> 
> A graphical application can do the animation as desired, because the concept of visual cursor and the concept of updating parts of the screen are independent. In terminals, updating the screen can pretty much only happen via the cursor, so the two concepts are heavily coupled.
> 
> Let's say, the user uses vim and presses a key that causes the cursor to jump from (1,1) to (5,7). What in fact happens behind the scenes is that the cursor moves to (60,24) (or wherever), the cursor position as it is printed in vim's status bar gets updated (moving the cursor forward through (61,24), (62,24) etc.), and finally it jumps to (5,7).
> 
> More complex changes might easily cause the cursor to paste through much larger portions of the screen, or even literally the entire screen.
> 
> If the cursor flashes up for a visual frame in one of these intermittent locations, it's probably not noticeable, or at least not distracting. If it's animated, even if simply from (1,1) to (60,24) and then from (60,24) to (5,7), it is a heavily noticeable undesired glitch – which is inevitable to occur every once in a while, due to the timing of events, network lag etc.

But that is true in the case of the attached gif as well, it is a
frontend for nvim, which means that the statusbar etc are all updated
via cursor movements. If this feature actually works there, it is likely
to work in a terminal as well.

But I agree that this would work best only if the program running in the
terminal was smart about its screen updates, using atomic updates (which
by the way, kitty does support), batching operations, etc.

**Comment by egmontkob** [comment]
> But that is true in the case of the attached gif as well, it is a frontend for nvim, which means that the statusbar etc are all updated via cursor movements.

I haven't studied the given application, but I have no idea how it could demonstrate animated cursor using its terminal version. Because then it either needs a patched terminal emulator, or a pretty full-featured terminal emulator as the frontend. I assume it's the graphical UI that has this feature, unrelated to any terminal business. Let me know if I'm mistaken :)

**Comment by kovidgoyal** [comment]
On Mon, Mar 23, 2020 at 04:55:15AM -0700, egmontkob wrote:
> > But that is true in the case of the attached gif as well, it is a frontend for nvim, which means that the statusbar etc are all updated via cursor movements.
> 
> I haven't studied the given application, but I have no idea how it could demonstrate animated cursor using its terminal version. Because then it either needs a patched terminal emulator, or a pretty full-featured terminal emulator as the frontend. I assume it's the graphical UI that has this feature, unrelated to any terminal business. Let me know if I'm mistaken :)

It is the graphical UI, but the graphical UIs for nvim work by running
an nvim instance in the background, and essentially passing the same
commands/data back and forth as would be done for a terminal. There are
of course some exceptions such as for graphical menus/popups that get
handles differently, but screen updates/cursor movemnt is as far as I
know powered by the same engine.

**Comment by egmontkob** [comment]
Wow, I didn't know this and never thought it would be done this way. Thanks for the info! :)

**Comment by keevan** [comment]
Just thought this might have been related to this discussion, since it discusses the animation, current _issues_ and workarounds
https://github.com/Kethku/neovide/issues/158

**Comment by kovidgoyal** [comment]
Since no one seems to want to do the work for this and I'm not a fan of the concept, closing. Will re-open if anyone shows interest in doing the work.

**Comment by Vaisakhkm2625** [comment]
i would love to have this feature... but i have no idea how opengl works.... and kitty source code felt really confusing for me due to use of multiple languages...  :( 
but i will try to learn and hopefully implement it :)...

**Comment by jas0xf** [comment]
according to https://www.reddit.com/r/vscode/comments/11e66xh/i_made_neovide_alike_cursor_effect_on_vscode/

he achieved the cursor animation by applying custom CSS from https://github.com/tholman/cursor-effects on cursor

is it possible to have css styling on cursor in kitty?

**Comment by LinkUpGames** [comment]
> Seems somewhat headache inducing to me, but am fine with an option for it that defaults to off. To implement it you would basically need to use the timers facility to implement the animation, similar to how it is done for blinking cursors.
> 
> If someone else wants to work on it feel free to post if you need help and I will provide assistance with getting started.

Hey! I'm willing to implement this feature. I saw a video of someone successfully implement a smooth cursor animation in Alacritty: https://www.reddit.com/r/unixporn/comments/1bvqnwr/oc_i_added_primitive_smooth_cursor_motion_to/

I was looking at the code and wondering where would be a good starting point. I noticed that the `child-monitor.c` file has some mentions of the blinking animation.  The function `collect_cursor_info()` directly uses the timers to calculate the opacity for the cursor. Am I in the right track?

**Comment by kovidgoyal** [comment]
Yes, that is correct. child-monitor is where the code decides if the cursor needs to be updated due to the blink animation. The cursor is actually drawn in shaders.c and the cell vertex and fragment shaders. However this machinery only supports drawing a cursor in a fixed cell position. What you will need to do is

1) First look at all the cursor movement commands in screen.c such as screen_cursor_position(). When such a command is received mark the cursor as being in "animated" state by adding a field to  the cursor struct. And store some animation data such as start position, and how much of animation is done, etc.

2) In child monitor handle cursors in the animated state specially with a timer, etc. This could should be fairly simple, kitty support CSS animation curves so use them. 

3) You will need to create a new shader and use that to render the cursor during the transition

4) Add an option to control this in oprions/definition.py and run gen/config.py

**Comment by mxple** [comment]
I'd like to implement this. My plan is to render the cursor the same way Neovide does it by keeping track of 4 corners of a pseudo-cursor which are constantly moving towards the real cursor. This means that the real cursor will not be rendered and that the fake cursor may not always be at the real cursor's location. This approach has the advantage of looking nicer (imo), but is more complex than just rendering a tail from the current location to the previous one.

There's a few concerns and questions I have going into this: 
1. How can/should I prevent the real cursor from being rendered based on a config value? Should I modify the shader, or pass in impossible values for the cursor location?
2. Essentially, I'm just rendering a quad in the cursor's color. However, I need to also cause text under the cursor to be a different color. I believe this must be done in 2 passes unless I make severe modifications to the `cell...` shaders (correct me if I'm wrong). Is a 2-pass approach acceptable performance-wise?
3. Are there any other predictable obstacles or caveats that I should know about?

**Comment by kovidgoyal** [comment]
On Thu, Sep 19, 2024 at 01:01:05PM -0700, Eric 李 wrote:
> I'd like to implement this. My plan is to render the cursor the same way Neovide does it by keeping track of 4 corners of a pseudo-cursor which are constantly moving towards the real cursor. This means that the real cursor will not be rendered and that the fake cursor may not always be at the real cursor's location. This approach has the advantage of looking nicer (imo), but is more complex than just rendering a tail from the current location to the previous one.
>
> There's a few concerns and questions I have going into this:
> 1. How can/should I prevent the real cursor from being rendered based on a config value? Should I modify the shader, or pass in impossible values for the cursor location?

The real cursor can already be hidden via a flag, either use it or add
another flag named "animating" or something similar and combine it with
the existing flag where the decision to render the cursor is made.

> 2. Essentially, I'm just rendering a quad in the cursor's color. However, I need to also cause text under the cursor to be a different color. I believe this must be done in 2 passes unless I make severe modifications to the `cell...` shaders (correct me if I'm wrong). Is a 2-pass approach acceptable performance-wise?

I dont follow, why would it need to be done in two passes? The current
cursor rendering code manages this in a single pass already. I dont
think you need to change the cell shaders at all, instead similar to the
animating flag above you add some data to the CursorRenderInfo struct
that changes where the cell shader renders the cursor. So no shader
changes are required.

Unless you mean its because the quad you are drawing is no longer cell
shaped? If that's the case then you would likely need to piggy back on
the infrastructure used for drawing images underneath text, and change
cell colors by actually changing the color in the cell structure just
while rendering and restoring it afterwards.

As for performance impact I dont care as long as that impact is
basically zero when the animation is disabled.

> 3. Are there any other predictable obstacles or caveats that I should know about?

Not that I can think of, note that the cell shader is undergoing some
refactoring currently in the trans branch so use that as a base for your
work if you are going to be touching the shader. The shader changes are
complete, some other work needs to be finished there before it is merged
back into master.

**Comment by mxple** [comment]
Thanks for the pointers!

> I dont follow, why would it need to be done in two passes? The current
cursor rendering code manages this in a single pass already. I dont
think you need to change the cell shaders at all, instead similar to the
animating flag above you add some data to the CursorRenderInfo struct
that changes where the cell shader renders the cursor. So no shader
changes are required.

> Unless you mean its because the quad you are drawing is no longer cell
shaped? If that's the case then you would likely need to piggy back on
the infrastructure used for drawing images underneath text, and change
cell colors by actually changing the color in the cell structure just
while rendering and restoring it afterwards.

Correct, the quad is arbitrarily shaped, and not aligned to any grid (see image). I'd like to make text under the quad the typical text-under-cursor color. but since the quad is arbitrarily shaped, the text within one cell make take on 2 colors. The other option is to not care about the text under an animated cursor, like in the image.
![image](https://github.com/user-attachments/assets/e3c0843c-4f95-4fd7-adde-8eea35c766b2)

Do you think I can use the existing image rendering infrastructure for this?

**Comment by kovidgoyal** [comment]
On Thu, Sep 19, 2024 at 10:27:33PM -0700, Eric 李 wrote:
> Thanks for the pointers!
>
> > I dont follow, why would it need to be done in two passes? The current
> cursor rendering code manages this in a single pass already. I dont
> think you need to change the cell shaders at all, instead similar to the
> animating flag above you add some data to the CursorRenderInfo struct
> that changes where the cell shader renders the cursor. So no shader
> changes are required.
>
> > Unless you mean its because the quad you are drawing is no longer cell
> shaped? If that's the case then you would likely need to piggy back on
> the infrastructure used for drawing images underneath text, and change
> cell colors by actually changing the color in the cell structure just
> while rendering and restoring it afterwards.
>
> Correct, the quad is arbitrarily shaped, and not aligned to any grid (see image). I'd like to make text under the quad the typical text-under-cursor color. but since the quad is arbitrarily shaped, the text within one cell make take on 2 colors. The other option is to not care about the text under an animated cursor, like in the image.
> ![image](https://github.com/user-attachments/assets/e3c0843c-4f95-4fd7-adde-8eea35c766b2)
>
> Do you think I can use the existing image rendering infrastructure for this?

I don't think its worth the effort, and I doubt anybody would care about
reading text under the cursor during an animation.

**Comment by Vaisakhkm2625** [comment]
in fact, mini.animate does a good job of animating cursor  in neovim without any support of terminal... https://github.com/echasnovski/mini.animate. 
only time i need animation is to screen share and teach someone something, which only occasionally happens... so def not worth the effort....

**Comment by jinhwanlazy** [comment]
Here is my initial implementation of the feature: https://github.com/jinhwanlazy/kitty/commit/689855bcd96ffa29f95c15f36449507064b516b2

https://github.com/user-attachments/assets/6b11b876-583f-410d-85aa-b75697c986fc

While there is still some work to be done, I believe it works well as intended. 

Known issues:
1. The options are not yet exposed 
1. The trail hides the cursor.
1. Cursor shapes and colors are not taken into consideration.
1. Behavior while resizing window is not taken into consideration.
1. Vim search bar steals cursor focus when sequential search next. (video 00:15, maybe it is neovim issue)

@kovidgoyal Could you take a look at my commit before I go further and make a PR?

**Comment by kovidgoyal** [comment]
Looks generally OK, I haven't reviewed in detail obviously will do that when you send a PR. One small comment is move the code to calculate the trail out of child-monitor either into its own module or cursors.c
